File: Practicas/Qytetet/QytetetRuby/lib/vista_textual_qytetet.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative "controlador_qytetet"
require_relative "opcion_menu"

module VistaTextualQytetet
  class VistaTextualQytetet
    attr_reader :controlador
    def initialize
      @controlador = ControladorQytetet::ControladorQytetet.instance
    end
    
    def obtener_nombre_jugadores
      nombres = Array.new
      puts("Introduzca el numero de jugadores que desee")
      numero_valido = false
      begin
        n = gets.chomp.to_i
        if(n>1 && n<10)
          numero_valido = true
        else
          puts("\nIntroduzca un numero de jugadores entre 2 y 10")
        end
      end while not numero_valido

      i = 0
      while(i<n)
        num_jugador = i+1
        puts("Introduzca el jugador " + num_jugador.to_s)
        s = gets.chomp
        nombres << s
        i+=1
      end
      nombres
    end

    def elegir_casilla(opcion_menu)
      casillas_validas = @controlador.obtener_casillas_validas(opcion_menu)
      if(casillas_validas.empty?)
        return -1
      else
        cadenas_casillas_validas = Array.new
        puts("\nCASILLAS VALIDAS (-1 para volver atras):")
        i = 0
        while(i < casillas_validas.length)
          opcion = casillas_validas.at(i).to_s
          cadenas_casillas_validas << opcion
          puts(opcion)
          i+=1
        end
        cadenas_casillas_validas << "-1" # Para poder volver atras
        opcion_valida = leer_valor_correcto(cadenas_casillas_validas).to_i
        opcion_valida
      end
    end

    def leer_valor_correcto(valores_correctos)
      begin
        puts("Introduzca una opcion: ")
        s = gets.chomp
      end while not valores_correctos.include?(s)
      s
    end

    def elegir_operacion
      operaciones_validas = @controlador.obtener_operaciones_juego_validas
      cadenas_operaciones_validas = Array.new
      puts("\nOPERACIONES:")
      i = 0
      while(i < operaciones_validas.length)
        opcion = operaciones_validas.at(i).to_s
        cadenas_operaciones_validas << opcion
        puts "#{opcion}: #{ControladorQytetet::OpcionMenu[opcion.to_i]}"
        i+=1
      end
      opcion_valida = leer_valor_correcto(cadenas_operaciones_validas).to_i
      opcion_valida
    end

    def self.main
      ui = VistaTextualQytetet.new
      ui.controlador.nombre_jugadores = ui.obtener_nombre_jugadores
      operacion_elegida = 0
      casilla_elegida = 0
      begin
        operacion_elegida = ui.elegir_operacion
        necesita_elegir_casilla = ui.controlador.necesita_elegir_casilla(operacion_elegida)
        if (necesita_elegir_casilla)
            casilla_elegida = ui.elegir_casilla(operacion_elegida)
        end
        if (!necesita_elegir_casilla || casilla_elegida >= 0)
            puts(ui.controlador.realizar_operacion(operacion_elegida,casilla_elegida))
        end
      end while(operacion_elegida != ControladorQytetet::OpcionMenu.index(:TERMINARJUEGO) &&
                operacion_elegida != ControladorQytetet::OpcionMenu.index(:OBTENERRANKING))
    end  
    VistaTextualQytetet.main
  end
end

File: Practicas/Qytetet/QytetetJava/src/vistatextualqytetet/VistaTextualQytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package vistatextualqytetet;
import controladorqytetet.*;
import java.util.ArrayList;
import java.util.Scanner;

/**
 *
 * @author enrique
 */
public class VistaTextualQytetet {
    private ControladorQytetet controlador;
    private static final Scanner in = new Scanner(System.in);
    
    public VistaTextualQytetet()
    {
        controlador = ControladorQytetet.getInstance();
    }
    
    public ArrayList<String> obtenerNombreJugadores() {
        
        ArrayList<String> nombres = new ArrayList<>();
        
        System.out.println("Introduzca el número de jugadores que desee");
        boolean numeroValido = false;
        int n;
        do
        {
            n = in.nextInt();
            if(n>1 && n<10)
                numeroValido = true;
            else
            {
                System.out.println("\nIntroduzca un número de jugadores entre 2 y 10");
            }
        } while(!numeroValido);
        

        for(int i = 0; i < n; ++i)
        {
            int numjugador = i+1;
            System.out.println("Introduzca el jugador " + numjugador);
            String s = in.next();
            nombres.add(s);
        }        
        return nombres; 
    }
    
    public int elegirCasilla(int opcionMenu)
    {
        ArrayList<Integer> casillasValidas = controlador.obtenerCasillasValidas(opcionMenu);
        if(casillasValidas.isEmpty())
            return -1;
        else
        {
            ArrayList<String> cadenasCasillasValidas = new ArrayList<>();
            System.out.println("\nCASILLAS VÁLIDAS (-1 para volver atrás):");
            for(int i=0; i < casillasValidas.size(); ++i)
            {
                String opcion = Integer.toString(casillasValidas.get(i));
                cadenasCasillasValidas.add(opcion);
                System.out.println(opcion);
            }
            cadenasCasillasValidas.add("-1"); // Para poder volver atrás
            int opcionValida = Integer.parseInt(leerValorCorrecto(cadenasCasillasValidas));
            return opcionValida;
        }
    }
    
    public String leerValorCorrecto(ArrayList<String> valoresCorrectos)
    {
        String s;
        do
        {
            System.out.println("Introduzca una opción: ");
            s = in.next();
        } while(!valoresCorrectos.contains(s));
        return s;
    }
    
    public int elegirOperacion()
    {
        ArrayList<Integer> operacionesValidas = controlador.obtenerOperacionesJuegoValidas();
        ArrayList<String> cadenasOperacionesValidas = new ArrayList<>();
        System.out.println("\nOPERACIONES:");
        for(int i=0; i < operacionesValidas.size(); ++i)
        {
            String opcion = Integer.toString(operacionesValidas.get(i));
            cadenasOperacionesValidas.add(opcion);
            System.out.println(opcion + ": " + OpcionMenu.values()[operacionesValidas.get(i)]);
        }
        int opcionValida = Integer.parseInt(leerValorCorrecto(cadenasOperacionesValidas));
        return opcionValida;
    }
    
    public static void main(String args[]) {
        VistaTextualQytetet ui = new VistaTextualQytetet();
        ui.controlador.setNombreJugadores(ui.obtenerNombreJugadores());
        int operacionElegida, casillaElegida = 0;
        boolean necesitaElegirCasilla;
        do {
            operacionElegida = ui.elegirOperacion();
            necesitaElegirCasilla = ui.controlador.necesitaElegirCasilla(operacionElegida);
            if (necesitaElegirCasilla)
                casillaElegida = ui.elegirCasilla(operacionElegida);
            if (!necesitaElegirCasilla || casillaElegida >= 0)
                System.out.println(ui.controlador.realizarOperacion(operacionElegida,casillaElegida));
        } while (operacionElegida != OpcionMenu.TERMINARJUEGO.ordinal() &&
                (operacionElegida != OpcionMenu.OBTENERRANKING.ordinal()));
    }
}
